Crash Protection Features

Hilimii opened the pull request:
Adds basic and advanced crash protection settings, adds ways to track them, and integrates them into `isLoading`.

Features to test:

- Basic crash protection setting. Pauses timer when the game closes only
- Advanced crash protection settings. Pauses timer when game is closed, or the player hasn't advanced beyond difficulty selection
- Settings for both added into Livesplit. Moved into own subheading outside Campaign for visibility
- Campaign Autostart option removed
- Adds a few variables to track crash protection
- Reduced refresh rate from 60 to 30 because 60 was firing so quickly that duplicate timings (two or more updates at the same moment) would cause logic issues

Hilimii committed:
Crash Protection Features

Adds basic and advanced crash protection settings, adds ways to track them, and integrates them into isLoading

jwfraustro commented:
Thanks for continuing to work on this! I'll take a look this evening, give it a test run and see if I find anything!

jwfraustro commented:
I'm unfortunately having trouble getting this to trigger correctly. I'll have more time tomorrow evening to look into it, it might be an issue on my end.

One thing to note about your troubles with `processUptime`  (nice address find, by the way!): rather than asking the autosplitter to compare the two floats every tick, which is expensive, it might be simpler to just use the `WingmanInstance`  address. It doesn't matter that it's actually an object, you can just set it as a byte and just watch that. Since we're just looking at the root address of the game instance, it should not change-- unless the game closes, at which point it will disappear. Doing some quick testing myself, that seems to work.

I'm still seeing some complaints in debugview about invalid memory handles, but I'll look into it a little bit more tomorrow.

Hilimii commented:
I've subbed in `WingmanInstance `instead of using `processUptime` for basic crash protection on my local draft and it works very well, good suggestion :)

Silly question, but are you comparing with game time in LiveSplit? I've accidentally turned it off a few times whilst clearing my splits and it caused `isLoading` to stop working.

jwfraustro commented:
Okay, I've had the chance to dig into this more and it is working for me now. I'm not sure what the exact issue was, but it was almost certainly on my end. 

I did some thinking about the two crash protection options. I think, for what the protection is trying to accomplish, only the "Advanced" protection is really necessary. 

The goal, while not officially supported as a speedrun category (yet?), is to preserve In-Game Times for Campaign speed runs. Consider the following scenarios where a person has some kind of crash protection on:

- A person is running IL's and crashes in-game:
They will most likely just reset the timer.
- A person is running campaign and crashes before the first mission loads:
They will most likely just reset the timer-- they will have to go through difficulty selection again, and there's no point in keeping the 1.2s before the crash.
- A person is running campaign and crashes at level 3:
We DO want to stop the timer and preserve it, and they will probably just press `Resume` when they relaunch, at which time it makes sense to restart the timer. 

I have put together a small demonstration video, cut for time, (pretend my resume was to some other mission than M01) that might illustrate what I mean. https://www.youtube.com/watch?v=d8l02-Xj10Y

I think it greatly cuts down the complexity of looking at `processUptime` (or the Instance address I suggested earlier). Additionally, if the category is ever considered, it becomes trivial to change the `isLoading` code to wait for the player to enter the hangar again (as you had suggested). Since the category is not currently supported it makes much of this moot, however, I hope you can see my point.

I think the behavior you have put in place for 'Advanced' is _exactly_ what a user would want to happen (and what might eventually be considered for IGT runs).

--- 

Two other small things:
- I think a tooltip for the crash protection combobox would be nice, since it's not immediately obvious what the option gives you.
- I'm still seeing the 'memory handle' error, but at this point I'm not concerned about it-- it hasn't affected anything I've done so far.

Hilimii commented:
I agree on all your points :) And the video looks good!

I think my reasoning behind creating basic crash protection was that it would appease people who would think that advanced protection would be unfaithful to an RTA category. If we're in agreement I'm happy to axe it tbh, from the perspective of the runner there's no reason they wouldn't want it to be advanced. Getting rid of basic allows us to put the crash setting back under the campaign mode setting too and make it look nicer + free up CPU usage, I'll give it a tooltip!

Also, yeah, I really don't know what the memory handle error is either and it hasn't caused issues so far 🤷‍♀️ so I'll let it be lol

I'll revise the code hopefully tonight, maybe tomorrow!

jwfraustro commented:
If/when we do get this merged in, we should probably increment to `1.1`. TBH, we probably should have done so with the takeoff skip, but I completely forgot about it.

Hilimii committed:
Cleaning and Slimlining Crash Settings

- Removes basic crash protection
- Advanced crash protection is now just crash protection
- Crash protection setting moved back to campaign mode
- References and usage of processUptime and WingmanInstance removed
- Update event no longer used
- Exit event no longer used
- Crashed variable removed
- isLoading crash protection now relies solely on onMissionSequence
- Cleaned up a few comments
- Added tooltips for all settings (I didn't know this was possible till it was mentioned! How exciting!)

Hilimii committed:
Update version Number to 1.1.0

jwfraustro commented:
Thanks for taking another pass at this! Skimming it, it all looks good. I just finished up the Kings/Faust splitting on my lunch break and I'll take a look at this after work.

jwfraustro commented:
This is great, and I'm very grateful for the new tooltips. Doing a quick test run saved me about 40s between getting back into the game, and back on the level sequence.
